src/use-swr.ts

```diff
@@ -488,6 +497,10 @@ export function useSWRHandler<Data = any, Error = any>(
     configRef.current = config
   })
 
+  useIsomorphicLayoutEffect(() => {
```

Comment:
This will cause the listeners being added by every `useSWR` hook. I think we should add the setup/teardown logic in `SWRConfig`.

Reply:
yes but we're allowed user to pass `cache` per hook, that would also be a inconsistent

Reply:
Oh you're right... :sadparrot: then maybe it's not a good idea to allow per hook cache?

Reply:
@shuding I think we need to figure out here what we want to support for global events revalidation, there will be ways to achieve : )
I'll call `config.isDocumentVisible` and `config.isOnline` as `env cond` (environment condition) in the following sections.


We have these combinations
* default cache + default env  ✅
* default cache + custom env❓
* custom cache + default env ❓
* custom cache + custom env ❓

If a hook has customized **env cond**, the global one will still do the routine. Each cache just need to be setup once with default **env cond** or customized **env cond**.

Support default **env cond**: WeakSet to count if the custom cache is existed, setup for it only once;
Support customized **env cond**: This only occurs in `SWRConfig`, setup once for each `SWRConfig`;

IMO, ideally the **env cond** won't change frequently, WeakSet can be an option cause you may have few `SWRConfig` with different cache but same **env cond**

Reply:
OK I think we should think about the most common use case: 99% apps only have 1 environment. So it's a good restriction that you shouldn't have per-hook, custom `isOnline` config, etc (they're safe to be singletons). 

Since we still haven't officially documented env settings, a good approach is to move those env settings into `createCache` (e.g. `createCache(provider, { isOnline, isDocumentVisible })`), so we can directly set up event listeners there too, before React's runtime.